x/gov/depinject.go

```diff
@@ -0,0 +1,139 @@
+package gov
+
+import (
+	"fmt"
+	"maps"
+	"slices"
+	"sort"
+	"strings"
+
+	modulev1 "cosmossdk.io/api/cosmos/gov/module/v1"
+	"cosmossdk.io/core/appmodule"
+	"cosmossdk.io/core/store"
+	"cosmossdk.io/depinject"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/codec"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/cosmos/cosmos-sdk/x/gov/keeper"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+	v1 "github.com/cosmos/cosmos-sdk/x/gov/types/v1"
+	"github.com/cosmos/cosmos-sdk/x/gov/types/v1beta1"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+)
+
+func init() {
+	appmodule.Register(
+		&modulev1.Module{},
+		appmodule.Provide(ProvideModule, ProvideKeyTable),
+		appmodule.Invoke(InvokeAddRoutes, InvokeSetHooks))
+}
+
+type ModuleInputs struct {
+	depinject.In
+
+	Config           *modulev1.Module
+	Cdc              codec.Codec
+	StoreService     store.KVStoreService
+	ModuleKey        depinject.OwnModuleKey
+	MsgServiceRouter baseapp.MessageRouter
+
+	AccountKeeper      govtypes.AccountKeeper
+	BankKeeper         govtypes.BankKeeper
+	StakingKeeper      govtypes.StakingKeeper
+	DistributionKeeper govtypes.DistributionKeeper
+
+	// CustomCalculateVoteResultsAndVotingPowerFn is an optional input to set a custom CalculateVoteResultsAndVotingPowerFn.
+	// If this function is not provided, the default function is used.
+	CustomCalculateVoteResultsAndVotingPowerFn keeper.CalculateVoteResultsAndVotingPowerFn `optional:"true"`
+
+	// LegacySubspace is used solely for migration of x/params managed parameters
+	LegacySubspace govtypes.ParamSubspace `optional:"true"`
+}
+
+type ModuleOutputs struct {
+	depinject.Out
+
+	Module       appmodule.AppModule
+	Keeper       *keeper.Keeper
+	HandlerRoute v1beta1.HandlerRoute
+}
+
+func ProvideModule(in ModuleInputs) ModuleOutputs {
+	defaultConfig := govtypes.DefaultConfig()
+	if in.Config.MaxMetadataLen != 0 {
+		defaultConfig.MaxMetadataLen = in.Config.MaxMetadataLen
+	}
+
+	// default to governance authority if not provided
+	authority := authtypes.NewModuleAddress(govtypes.ModuleName)
+	if in.Config.Authority != "" {
+		authority = authtypes.NewModuleAddressOrBech32Address(in.Config.Authority)
+	}
+
+	k := keeper.NewKeeper(
+		in.Cdc,
+		in.StoreService,
+		in.AccountKeeper,
+		in.BankKeeper,
+		in.StakingKeeper,
+		in.DistributionKeeper,
+		in.MsgServiceRouter,
+		defaultConfig,
+		authority.String(),
+	)
+	m := NewAppModule(in.Cdc, k, in.AccountKeeper, in.BankKeeper, in.LegacySubspace)
+	hr := v1beta1.HandlerRoute{Handler: v1beta1.ProposalHandler, RouteKey: govtypes.RouterKey}
+
+	if in.CustomCalculateVoteResultsAndVotingPowerFn != nil {
+		k.SetCustomCalculateVoteResultsAndVotingPowerFn(in.CustomCalculateVoteResultsAndVotingPowerFn)
+	}
```

Comment:
thoughts on using an options pattern in the keeper constructor? we could set the default on the keeper and then run opts, that way we can always expect it to be set.